### ISSUE TITLE
Save optimizer state?

### ISSUE THREAD

**Issue by ericphanson** [beacon-biosignals/LegolasFlux.jl#7]
This may not need any code changes, but it would be great to update the example to save out optimizer state to do resumable training. This might require waiting for https://github.com/FluxML/Optimisers.jl to mature a bit.

Or else we need to do some more complicated stuff to handle things like `IdDict`'s in `Flux.ADAM`. E.g. we could map `objectid`'s to something like `UUIDs`, then save those out next to the arrays, e.g. our `weights` object could hold a `Vector{Tuple{UUID, FlatArray{T}}}` instead of `Vector{FlatArray{T}}`. Then we could write code to map between `IdDict`s and another `Vector{Tuple{UUID, FlatArray{T}}}` (or maybe just `Dict{UUID, FlatArray{T}}` if Arrow will support that easily) to save out the optimizer internals.

Something like that would be needed if we wanted to support having the same layer show up twice in a model with identical weights (I think this could come up in weight-sharing although not all forms of weight-sharing would need this support).

**Comment by haberdashPI** [comment]
This would be nice. I don't know much about `Optimisers.jl` but with Flux.Optimisers you can't use `cpu` to pull out the state vector (or put it back with `gpu`) if a model is run on the GPU (it does not change the arrays in the `IdDict`). There is a little extra wrangling that would need to happen. I think it could make sense to have some struct that you pass both the model parameters and the optimiser to that you can then define methods for `fetch_weights` and `load_weights!` on. That would make an implementation that deals with these annoying bits pretty feasible.

**Comment by ericphanson** [comment]
Cool, that makes sense. We might be able to facilitate passing models to the gpu and back since we can't use the `objectid` for serialization anyway (since it doesn't persist between sessions), so however we manage to serialize them (e.g. assigning UUIDs or just array indices to weights) we can use that to associate momentum's etc (what's in the `IdDict`) to those arrays even as we move the arrays around and change their `objectid`s. Not 100% sure how that API would work but we'd at least have the info needed, since we need that for serialization anyway.

**Comment by ericphanson** [comment]
Another thing we could do is instead of trying to handle arbitrary optimizers, we could just copy over Flux's ADAM optimizer and modify it to not use an `IdDict` (perhaps just a flat vector of the same size as the `weights` vector we keep). Ideally LegolasFlux wouldn't be in the business of making optimizers but it could be a quick way to get something working for now.

**Comment by haberdashPI** [comment]
Yeah, I think the `IdDict` thing is the main issue. One way to cheat would be to rely on the fact that a pass through something like `fcollect` will yield the same ordering, and so you can fcollect with the original arrays and the new, converted arrays to find the correspondence. This would make the implementation sensitive to internal changes: e.g. re-ordering fields of a struct could change how fcollect orders things (but that was an issue anyways).
